server/public/plugin/hooks.go

```diff
@@ -336,6 +346,96 @@ type Hooks interface {
 	// Minimum server version: 9.2
 	ServeMetrics(c *Context, w http.ResponseWriter, r *http.Request)
 
+	// UserWillJoinChannel is invoked before a user joins a channel.
+	//
+	// To reject a user joining a channel, return an non-empty string describing why the user was rejected.
+	// To allow the user to join the channel, return an empty string.
+	//
+	// Minimum server version: 9.6
+	// Requires enterprise license
+	UserWillJoinChannel(c *Context, user *model.User, channel *model.Channel, requester *model.User) string
+
+	// UserWillLeaveChannel is invoked before a user leaves a channel.
+	//
+	// To reject a user leaving a channel, return an non-empty string describing why the user was rejected.
+	// To allow the user to leave the channel, return an empty string.
+	//
+	// Minimum server version: 9.6
+	// Requires enterprise license
+	UserWillLeaveChannel(c *Context, channelMember *model.ChannelMember) string
+
+	// UserWillJoinTeam is invoked before a user joins a team.
+	//
+	// To reject a user joining a team, return an non-empty string describing why the user was rejected.
+	// To allow the user to join the team, return an empty string.
+	//
+	// Minimum server version: 9.6
+	// Requires enterprise license
+	UserWillJoinTeam(c *Context, user *model.User, team *model.Team) string
+
+	// UserWillLeaveTeam is invoked before a user leaves a team.
+	//
+	// To reject a user leaving a team, return an non-empty string describing why the user was rejected.
+	// To allow the user to leave the team, return an empty string.
+	//
+	// Minimum server version: 9.6
+	// Requires enterprise license
+	UserWillLeaveTeam(c *Context, teamMember *model.TeamMember) string
+
+	// UserWillBeCreated is invoked before a user is created.
+	//
+	// To reject a user creation, return an non-empty string describing why the user was rejected.
+	// To allow the user to be created, return an empty string.
+	//
+	// Minimum server version: 9.6
+	// Requires enterprise license
+	UserWillBeCreated(c *Context, user *model.User) string
```

Comment:
For other hooks like `FileWillBeUploaded` the plugin developer is allowed to modify the entity. While I'm not sure if there are use cases for the hooks like `UserWillJoinTeam`, I could see ones for `UserWillBeCreated`, e.g. change the user name to fit a format or auto-set the profile image. Should we change the signature to 
```suggestion
	UserWillBeCreated(c *Context, user *model.User) (*model.User, string)
```
?

Reply:
Given how hard it is to make changes to the plugin API, I wonder if we should add the option to change the entity to all `WillBe` hooks. It might add additional value to them.

Reply:
That is a really good point, I love it. I would try to add it to all the new ones, unless clearly makes no sense, but I don't see any reason to not do it, in general.